### PR TITLE
tlsutil: Fix check TLS configuration

### PR DESCRIPTION
**Pull request by beautifulentropy**
### Description

Since the introduction of #5394, the `OutgoingTLSConfigForCheck` method of `tlsutil.Configurator` has been setting the ServerName field of TLS configurations used for executing health checks to the Consul node or server name by default.

According to the crypto/tls documentation:

    // ServerName is used to verify the hostname on the returned
    // certificates unless InsecureSkipVerify is given. It is also included
    // in the client's handshake to support virtual hosting unless it is
    // an IP address.

Unless the `check.tls_server_name` indicates otherwise, it is generally better to leave the `ServerName` field empty. The crypto/tls client will infer the ServerName for SNI from the check address, unless it's an IP (as per RFC 6066, Section 3). There are two common situations where specifying this field can be beneficial:

1. When the check address is an IP, `tls_server_name` can be specified for SNI. Note: setting `tls_server_name` will also override the hostname used to verify the certificate presented by the server being checked.

2. When the hostname in the check address won't be present in the SAN (Subject Alternative Name) field of the certificate presented by the server being checked. Note: setting `tls_server_name` will also override the hostname used for SNI.

This PR also updates the documentation for `tls_server_name` and `tls_skip_verify` to indicate that both fields can be used for H2Ping and gRPC checks, not just HTTP and provides a slightly slimmed down version of the above circumstances for `tls_server_name`.

### Testing & Reproduction steps

I encountered this issue while drafting letsencrypt/boulder#6908. After experiencing certificate validation failures due the Consul node name failing to be present in the SAN field of certificate of the component being health checked, we dug into the Consul code and diagnosed the issue above.

### Links

Substantiating my claim that crypto/tls client will infer the ServerName based on the address being dialed: [link](https://cs.opensource.google/go/go/+/refs/tags/go1.20.4:src/crypto/tls/tls.go;l=147-154)

### PR Checklist

* [X] updated test coverage
* [X] external facing docs updated
* [ ] appropriate backport labels added
* [ ] not a security concern

### CODE REVIEW

**Comment by beautifulentropy** [comment]
@blake, this PR has been open for a while. Is there someone in particular that I should ping to get a reviewer assigned?

File: website/content/docs/services/configuration/checks-configuration-reference.mdx [hunk 1]
```diff
@@ -35,8 +35,8 @@ Specify health check options in the `check` block. To register two or more heath
...
+| `tls_skip_verify` | Boolean value that disables verification of the chain and hostname of the certificate presented by the server being checked. This is generally not suitable for production use. Default is `false`. | <li>HTTP </li> <li>H2Ping </li> <li>gRPC </li> |
```
Comment (trujillo-adam) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:39]:
```suggestion
| `tls_skip_verify` | Boolean value that determines if the check verifies the chain and hostname of the certificate that the server presents. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. | <li>HTTP </li> <li>H2Ping </li> <li>gRPC </li> |
```
Switch to active voice.

**Review by trujillo-adam — APPROVED** [review]
Sorry for the long delay on reviewing the docs. I made a suggestion and am pre-approving so that you are not blocked.

**Comment by david-yu** [comment]
Thanks @beautifulentropy this change makes sense. This change likely will need to go into a future minor release on Consul instead of a patch release since it is a breaking change.

**Comment by beautifulentropy** [comment]
> Thanks @beautifulentropy this change makes sense. This change likely will need to go into a future minor release on Consul instead of a patch release since it is a breaking change.

This is exactly what I figured. It's a super small change with a fairly big blast radius for folks who aren't presently using `tls_skip_verify` on all of their gRPC checks.

**Review by david-yu — APPROVED** [review]
LGTM but we'll need someone from Eng to also take a pass and review this as well. This will likely  land in 1.17.0

File: tlsutil/config.go [hunk 2]
```diff
@@ -857,10 +857,23 @@ func (c *Configurator) IncomingHTTPSConfig() *tls.Config {
...
+// It is RECOMMENDED that the serverName be left unspecified. The crypto/tls
+// client will deduce the ServerName (for SNI) from the check address unless
+// it's an IP (RFC 6066, Section 3). However, there are two instances where
+// supplying a serverName is useful:
+//
+//  1. When the check address is an IP, a serverName can be supplied for SNI.
+//     Note: setting serverName will also override the hostname used to verify
+//     the certificate presented by the server being checked.
+//
+//  2. When the a hostname in the check address won't be present in the SAN
```
Comment (wangxinyi7) [inline tlsutil/config.go:870]: `the a`? You mean `When the hostname in the check` right?

File: website/content/docs/services/configuration/checks-configuration-reference.mdx [hunk 3]
```diff
@@ -35,8 +35,8 @@ Specify health check options in the `check` block. To register two or more heath
...
-| `tls_skip_verify` | Boolean value that disbles TLS for HTTP checks when set to `true`. Default is `false`. | <li>HTTP</li> |
+| `tls_server_name` | String value that specifies the server name used to verify the hostname on the returned certificates unless `tls_skip_verify` is given. Also included in the client's handshake to support SNI. It is recommended that this field be left unspecified. The TLS client will deduce the server name for SNI from the check address unless it's an IP ([RFC 6066, Section 3](https://tools.ietf.org/html/rfc6066#section-3)). There are two common circumstances where supplying a `tls_server_name` can be beneficial: <li>When the check address is an IP, `tls_server_name` can be specified for SNI. Note: setting `tls_server_name` will also override the hostname used to verify the certificate presented by the server being checked.</li><li>When the a hostname in the check address won't be present in the SAN (Subject Alternative Name) field of the certificate presented by the server being checked. Note: setting `tls_server_name` will also override the hostname used for SNI.</li> | <li>HTTP </li> <li>H2Ping </li> <li>gRPC </li> |
```
Comment (wangxinyi7) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:38]: `the a`

**Comment by wangxinyi7** [comment]
@beautifulentropy Thanks for putting things together! I left very small comments but nothing is blocking. Approved.

**Review by wangxinyi7 — APPROVED** [review]

**Comment by david-yu** [comment]
Could you also add a `.changelog/17481.txt` with a changelog file that looks like the following? 

```release-note:breaking-change
 tlsutil: ... 
 ```

**Comment by beautifulentropy** [comment]
Thanks for the comments, I'll take a look at this tomorrow.

File: tlsutil/config.go (hunk 2 above)
Comment (beautifulentropy) [reply tlsutil/config.go:870]: Good catch!

**Comment by david-yu** [comment]
Thanks @beautifulentropy. We reviewed the PR and it looks great. Will go ahead and merge. This will be introduced in Consul 1.17.0 as a breaking change. Thank you.

**Pull request merged**